### PR TITLE
ensure permissions for potential duplicate group names across IdPs/te…

### PR DESCRIPTION
**Pull request by bsmithvt**
…nants are resolved correctly

### CODE REVIEW

**Comment by ryanorbaker** [comment]
for my own understanding:  if I understand the change, we are switching from passing group names around to passing group objects (with names and ids) so we can duplicate names?

File: Fabric.Authorization.Domain/Resolvers/Permissions/RolePermissionResolverService.cs [hunk 1]
```diff
@@ -78,14 +78,16 @@ public async Task<PermissionResolutionResult> Resolve(PermissionResolutionReques
...
-        private static bool BypassRoleEvaluation(Role role, IReadOnlyCollection<string> groupNames, PermissionResolutionRequest permissionResolutionRequest)
+        private static bool BypassRoleEvaluation(Role role, IReadOnlyCollection<Group> groups, PermissionResolutionRequest permissionResolutionRequest)
         {
-            return !(role.Groups.Any(g => groupNames.Contains(g.Name, StringComparer.OrdinalIgnoreCase)) || role.Users.Any(
-                         u => u.SubjectId.Equals(permissionResolutionRequest.SubjectId,
-                                  StringComparison.OrdinalIgnoreCase) &&
-                              u.IdentityProvider.Equals(permissionResolutionRequest.IdentityProvider,
-                                  StringComparison.OrdinalIgnoreCase))) || role.IsDeleted ||
-                   role.Permissions == null || role.Permissions.Count == 0;
+            return !(role.Groups.Any(g => groups.Select(g1 => g1.GroupIdentifier).Contains(g.GroupIdentifier, new GroupIdentifierComparer()))
```
Comment (ryanorbaker) [inline Fabric.Authorization.Domain/Resolvers/Permissions/RolePermissionResolverService.cs]: is this where you think we need to use the *correct* ID, either from Azure or other?

**Comment by bsmithvt** [comment]
@ryanorbaker, yes, since with Azure AD we can have duplicate group names that span tenants, the new group identifier will be the IdP + tenant + group name.

File: Fabric.Authorization.API/Modules/UsersModule.cs [hunk 2]
```diff
@@ -134,7 +134,7 @@ private async Task<dynamic> GetUserPermissions(dynamic param)
                 Grain = userPermissionRequest.Grain,
                 SecurableItem = userPermissionRequest.SecurableItem,
                 IncludeSharedPermissions = isGrainEmpty,
-                UserGroups = groups?.Select(g => g.Name)
+                UserGroups = groups
```
Comment (michaelvidal24) [inline Fabric.Authorization.API/Modules/UsersModule.cs]: Do we need a similar change here: https://github.com/HealthCatalyst/Fabric.Authorization/pull/389/files#diff-8aa3fecd22b341d99ca92ccc12c0f59dR159, or is that already handled?

File: Fabric.Authorization.Domain/Resolvers/Permissions/RolePermissionResolverService.cs (hunk 1 above)
Comment (bsmithvt) [reply Fabric.Authorization.Domain/Resolvers/Permissions/RolePermissionResolverService.cs]: The GroupIdentifier is a property consisting of IdP + tenant ID + group name. This logic was switched from comparing group names to comparing `GroupIdentifier` objects.

Comment (bsmithvt) [reply Fabric.Authorization.Domain/Resolvers/Permissions/RolePermissionResolverService.cs]: The Azure identifier is used to correlate internal Fabric.Auth groups to Azure groups. We correlate through the new `ExternalIdentifier` column. Access tokens from Azure AD will have that in the `groups` claim so we have to translate internally. This is done up the call stack from this method.

File: Fabric.Authorization.API/Modules/UsersModule.cs (hunk 2 above)
Comment (bsmithvt) [reply Fabric.Authorization.API/Modules/UsersModule.cs]: Not sure what you mean by "similar change", but this endpoint should be handled. I switched the `PermissionResolutionRequest` from having a list of group names to having a list of `GroupIdentifier` objects.

Comment (michaelvidal24) [reply Fabric.Authorization.API/Modules/UsersModule.cs]: Just wanting to make sure the GetCurrentUserPermissions has been updated to reflect UserGroups changing from being a list of `string` to a list of `GroupIdentifier` objects.

Comment (bsmithvt) [reply Fabric.Authorization.API/Modules/UsersModule.cs]: Ah yeah, it has been.

**Pull request merged**